plugins/destination/postgresql/client/insert.go

```diff
@@ -59,27 +87,61 @@ func (c *Client) InsertBatch(ctx context.Context, messages message.WriteInserts)
 			return fmt.Errorf("table %s not found", tableName)
 		}
 
-		sql, ok := queries[tableName]
+		detail, ok := details[tableName]
 		if !ok {
+			var sql string
 			// cache the query
 			table := c.normalizeTable(msg.GetTable())
 			if len(table.PrimaryKeysIndexes()) > 0 {
 				sql = c.upsert(table)
 			} else {
 				sql = c.insert(table)
 			}
-			queries[tableName] = sql
+			detail = tableDetail{
+				indexes: table.PrimaryKeysIndexes(),
+				sql:     sql,
+				values:  make(map[string]bool),
```

Comment:
```suggestion
				values:  make(map[string]struct{}),
```

using empty `struct{}` here should save a bit of memory